### PR TITLE
Add CIDR overlap verification.

### LINKED ISSUE #535: Shoot networks must not overlap when multi AZ is used
Currently Gardener could deploy `Shoot` cluster across multiple availability zones.
For this purpose in the shoot manifest every zone must  has its own `internal` CIDR, `public` CIDR and `workers` CIDR, but a proper validation is missing and Gardener will not disallow shoot with overlapped networks.

### PR DESCRIPTION
**Pull request by mvladev**
**What this PR does / why we need it**:
At this moment we lack verification for when different CIDRs must or not overlap. Users can create `Seed` and `Shoot` resources which simply fail to be created.

On all providers `Pod`, `Nodes` and `Services` CIDRs must not overlap.
Following provider-specific verification are added:

AWS:

- `Worker`, `Public` and `Internal`, `Pods` and `Services` must not overlap.
- `Nodes` must contain all of `Worker` CIDRs.
- If VPC `cird` is provovided, then it must contain `Node`,  `Public`, `Internal` and `Workers` CIDRs.
- If VPC `cird` is provovided, then it must not contain `Pods`  and `Services` CIDRs.

Alicloud:

- `Worker`, `Pods` and `Services` must not overlap.
- `Nodes` must contain all of `Worker` CIDRs.
- If VPC `cird` is provovided, then it must contain `Node`  and `Workers` CIDRs.
- If VPC `cird` is provovided, then it must not contain `Pods`  and `Services` CIDRs.

Azure:

- `Worker`, `Pods` and `Services` must not overlap.
- `Nodes` must contain all of `Worker` CIDRs.
- If Vnet `cird` is provovided, then it must contain `Node`  and `Workers` CIDRs.
- If Vnet `cird` is provovided, then it must not contain `Pods`  and `Services` CIDRs.

GCP:

- `Worker`, `Pods` and `Services` must not overlap.
- `Nodes` must contain all of `Worker` CIDRs.

OpenStack:

- `Worker`, `Pods` and `Services` must not overlap.
- `Nodes` must contain all of `Worker` CIDRs.

**Which issue(s) this PR fixes**:
Fixes #535 

**Special notes for your reviewer**:
Please take a look at the release note as it's a backwards incompatible change (as is any verification changes).

I'll update the shoot validator in separate PR.

**Release note**:

``` action operator
⚠️ Strict `Seed` and `Shoot` CIDR verification rules are now enabled.
`Shoot` or `Seed` resource created previously with invalid configurations MUST be deleted, before updating to this version. Failing to do so, will cause ANY modification to said resources (including attempted deletion) to be rejected by the apiserver.
```

### CODE REVIEW

**Comment by timuthy** [comment]
This validation is very helpful for users. I question whether it's that critical to introduce such a check at the moment, though.

IMO, we'd rather push our **Extensibility** plans and this change contradicts since it adds cloud specific logic to an area that will be supposed to become cloud neutral.

**Comment by mvladev** [comment]
@timuthy I agree, but even with extensibility, those checks should be added to all providers anyways. It's better to have them, so they can be ported to external providers.

**Review by rfranzke — COMMENTED** [review]
Agreed with @timuthy, this is a huge change and hard to review...
Anyway, I went through it and tested it locally - it seems to be OK.

Please enhance the tests, so use

```golang
Expect(errorList).To(ConsistOf(
	PointTo(MatchFields(IgnoreExtras, Fields{
		"Type":  Equal(field.ErrorTypeInvalid),
		"Field": Equal("spec.cloud.profile"),
})),
````

instead of e.g.,

```golang
Expect(el).To(HaveLen(1))
Expect(*el[0]).To(Equal(field.Error{
	Type:     field.ErrorTypeInvalid,
	Field:    badPath.String(),
	BadValue: badCIDR,
	Detail:   `must not be a subset of "foo" ("10.0.0.0/8")`,
}))
```

or your custom function

```golang
assertErrorList(...)
```

**Comment by mvladev** [comment]
> Please enhance the tests

@rfranzke I avoided that, so the PR doesn't become even further bloated.

**Comment by rfranzke** [comment]
Please do it at least for your newly introduced tests.

**Comment by mvladev** [comment]
@rfranzke which tests are you referring to? All of them are using this new `assertErrorList(...)`

**Comment by rfranzke** [comment]
See above, all new tests you introduced that either use

```golang
Expect(el).To(HaveLen(1))
Expect(*el[0]).To(Equal(field.Error{
	Type:     field.ErrorTypeInvalid,
	Field:    badPath.String(),
	BadValue: badCIDR,
	Detail:   `must not be a subset of "foo" ("10.0.0.0/8")`,
}))
```

or your custom function

```golang
assertErrorList(...)
```

**Comment by mvladev** [comment]
@rfranzke I've added several matcher functions for composition and simplification

```go
func HaveFields(fields Fields) gomegatypes.GomegaMatcher {
	return PointTo(MatchFields(IgnoreExtras, fields))
}

func ConsistOfFields(fields ...Fields) gomegatypes.GomegaMatcher {
	m := []interface{}{}
	for _, f := range fields {
		m = append(m, HaveFields(f))
	}
	return ConsistOf(m...)
}
```

with this setup instead of 

```go
Expect(errorList).To(ConsistOf(
    PointTo(MatchFields(IgnoreExtras, Fields{
        "Type":  Equal(field.ErrorTypeInvalid),
        "Field": Equal("bar"),
    })),
    PointTo(MatchFields(IgnoreExtras, Fields{
        "Type":  Equal(field.ErrorTypeInvalid),
        "Field": Equal("baz"),
    })),
))

```

you get the simplified solution

```go
Expect(errorList).To(ConsistOfFields(
    Fields{
        "Type":  Equal(field.ErrorTypeInvalid),
        "Field": Equal("bar"),
    }, Fields{
        "Type":  Equal(field.ErrorTypeInvalid),
        "Field": Equal("baz"),
    },
))
```

**Comment by mvladev** [comment]
@rfranzke I've fixed the test issues in 5c5c498 for easy review . I'll squish it before merge.

File: pkg/utils/validation/cidr/cidr_test.go [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+
+		It("should return a nil FieldPath", func() {
+			cdr := NewCIDR(validGardenCIDR, path)
+			badCIDR := garden.CIDR("10.0.0.1/32")
+			badPath := field.NewPath("bad")
+			other := NewCIDR(badCIDR, badPath)
+
+			el := cdr.ValidateNotSubset(other)
+			Expect(el).To(HaveLen(1))
+			Expect(*el[0]).To(Equal(field.Error{
```
Comment (rfranzke) [inline pkg/utils/validation/cidr/cidr_test.go]: Not yet changed

File: pkg/utils/validation/cidr/cidr_test.go [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+			cdr := NewCIDR(validGardenCIDR, path)
+
+			Expect(cdr.ValidateParse()).To(BeEmpty())
+		})
+
+		It("should return a nil FieldPath", func() {
+			cdr := NewCIDR(invalidGardenCIDR, path)
+
+			el := cdr.ValidateParse()
+			Expect(el).To(HaveLen(1))
```
Comment (rfranzke) [inline pkg/utils/validation/cidr/cidr_test.go]: Not yet changed

File: pkg/utils/validation/cidr/cidr_test.go [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+			Expect(cdr.ValidateSubset(other)).To(BeEmpty())
+		})
+
+		It("should not be a subset", func() {
+			cdr := NewCIDR(validGardenCIDR, path)
+			other := NewCIDR(garden.CIDR("10.0.0.1/32"), field.NewPath("bad"))
+
+			el := other.ValidateSubset(cdr)
+			Expect(el).To(HaveLen(1))
+			Expect(*el[0]).To(Equal(field.Error{
```
Comment (rfranzke) [inline pkg/utils/validation/cidr/cidr_test.go]: Not yet changed

**Comment by mvladev** [comment]
test issues resolved in 89bbc29

**Review by rfranzke — APPROVED** [review]
/lgtm

**Comment by rfranzke** [comment]
@mvladev Can you please check whether correctness of the provided CIDRs is validated?
We have seen a cluster that used `10.250.0.96/26` which is invalid.

To give more background information: A user was trying to create a multi-zone AWS cluster with those three CIDRs:

```
10.250.0.0/26
10.250.0.96/26
10.250.0.128/26
```

Not only was the second one wrong but also it would overlap with the third one. Can you check if we have a bug, please? https://sap-cp.slack.com/archives/C9CEBQPGE/p1566475961208700

**Comment by mvladev** [comment]
> We have seen a cluster that used 10.250.0.96/26 which is invalid.

It's a valid CIDR. See https://play.golang.com/p/B8TfNEaH84h. The netmask is `255.255.255.192` and it applies for addresses `10.250.0.64` - `10.250.0.127`.

Should it be specified as the canonical  `10.250.0.64/26`? Probably. Is it invalid CIDR - it's not.

> Not only was the second one wrong but also it would overlap with the third one.

There is no overlap:

```bash
$ sipcalc 10.250.0.0/26
-[ipv4 : 10.250.0.0/26] - 0

[CIDR]
Host address            - 10.250.0.0
Host address (decimal)  - 184156160
Host address (hex)      - AFA0000
Network address         - 10.250.0.0
Network mask            - 255.255.255.192
Network mask (bits)     - 26
Network mask (hex)      - FFFFFFC0
Broadcast address       - 10.250.0.63
Cisco wildcard          - 0.0.0.63
Addresses in network    - 64
Network range           - 10.250.0.0 - 10.250.0.63
Usable range            - 10.250.0.1 - 10.250.0.62

$ sipcalc 10.250.0.96/26
-[ipv4 : 10.250.0.96/26] - 0

[CIDR]
Host address            - 10.250.0.96
Host address (decimal)  - 184156256
Host address (hex)      - AFA0060
Network address         - 10.250.0.64
Network mask            - 255.255.255.192
Network mask (bits)     - 26
Network mask (hex)      - FFFFFFC0
Broadcast address       - 10.250.0.127
Cisco wildcard          - 0.0.0.63
Addresses in network    - 64
Network range           - 10.250.0.64 - 10.250.0.127
Usable range            - 10.250.0.65 - 10.250.0.126

$ sipcalc 10.250.0.128/26
-[ipv4 : 10.250.0.128/26] - 0

[CIDR]
Host address            - 10.250.0.128
Host address (decimal)  - 184156288
Host address (hex)      - AFA0080
Network address         - 10.250.0.128
Network mask            - 255.255.255.192
Network mask (bits)     - 26
Network mask (hex)      - FFFFFFC0
Broadcast address       - 10.250.0.191
Cisco wildcard          - 0.0.0.63
Addresses in network    - 64
Network range           - 10.250.0.128 - 10.250.0.191
Usable range            - 10.250.0.129 - 10.250.0.190
````

**Comment by rfranzke** [comment]
OK, can we then implement validation that only canonical CIDRs are used?

**Pull request merged**